This patch is wrong, only a remedy.
The issue is that both e2fsprogs and heimdal install compile_et, the one
being installed depends on the compile sequence.
The one needed for samba is from heimdal-host.
When source4/heimdal/lib/asn1/asn1_err.c is generated by the wrong compile_et, 'link' is defined and conflicting with 'link' from unistd.h.

A comment in package/heimdal/heimdal.mk:
# Don't use compile_et from e2fsprogs as it raises a build failure with samba4

--- samba4-4.15.7/lib/replace/replace.h-orig	2023-08-26 18:50:30.243506358 +0100
+++ samba4-4.15.7/lib/replace/replace.h	2023-08-26 18:51:10.270762843 +0100
@@ -168,10 +168,6 @@
 #include <bsd/unistd.h>
 #endif
 
-#ifdef HAVE_UNISTD_H
-#include <unistd.h>
-#endif
-
 #ifdef HAVE_STRING_H
 #include <string.h>
 #endif
